File: typescript/packages/core/tests/sensor.test.ts
/**
 * Unit tests for the canonical SensorChain (spore-core issue #10).
 *
 * Mirrors `rust/crates/spore-core/src/sensor.rs#tests` rule-for-rule.
 */

import { describe, expect, it } from "vitest";

import { sensor, SessionId } from "../src/index.js";
import { emptySessionState } from "../src/harness/types.js";

const { StandardSensorChain, SensorId, Timestamp, SensorError, newSensorInput } = sensor;

type Sensor = sensor.Sensor;
type SensorConfig = sensor.SensorConfig;
type SensorInput = sensor.SensorInput;
type SensorOutcome = sensor.SensorOutcome;
type SensorTrigger = sensor.SensorTrigger;
type SensorResult = sensor.SensorResult;

// ── Builders ────────────────────────────────────────────────────────────────

function ts(s: string): sensor.Timestamp {
  return new Timestamp(s);
}

class StubSensor implements Sensor {
  constructor(
    private readonly cfg: SensorConfig,
    private readonly outcome: SensorOutcome,
  ) {}
  async evaluate(_input: SensorInput): Promise<SensorResult> {
    return {
      sensor_id: new SensorId(this.cfg.id.value),
      outcome: this.outcome,
      observation: this.outcome === "warn" ? "warn-obs" : null,
      detail: this.outcome,
      fired_at: ts("2026-05-16T00:00:00Z"),
    };
  }
  config(): SensorConfig {
    return this.cfg;
  }
}

function computational(
  id: string,
  triggers: SensorTrigger[],
  outcome: SensorOutcome,
  overrides: Partial<SensorConfig> = {},
): StubSensor {
  return new StubSensor(
    {
      id: new SensorId(id),
      name: id,
      kind: "computational",
      triggers,
      run_every_n_turns: null,
      run_on_phases: null,
      low_signal_threshold: sensor.defaultSensorSignalThresholds(),
      ...overrides,
    },
    outcome,
  );
}

function inferential(
  id: string,
  triggers: SensorTrigger[],
  outcome: SensorOutcome,
  every_n: number | null,
  phases: sensor.SensorConfig["run_on_phases"],
): StubSensor {
  return new StubSensor(
    {
      id: new SensorId(id),
      name: id,
      kind: "inferential",
      triggers,
      run_every_n_turns: every_n,
      run_on_phases: phases,
      low_signal_threshold: sensor.defaultSensorSignalThresholds(),
    },
    outcome,
  );
}

function input(sid: string): SensorInput {
  return newSensorInput(new SessionId(sid), emptySessionState());
}

// ── Tests ───────────────────────────────────────────────────────────────────

describe("SensorChain — register validation", () => {
  it("register_rejects_empty_triggers", () => {
    const chain = new StandardSensorChain();
    const s = computational("s1", [], "pass");
    expect(() => chain.register(s)).toThrow(SensorError);
    try {
      chain.register(s);
    } catch (e) {
      expect((e as sensor.SensorError).kind).toBe("validation_failed");
    }
  });

  it("register_rejects_duplicate_ids", () => {
    const chain = new StandardSensorChain();
    chain.register(computational("s1", [{ kind: "post_turn" }], "pass"));
    try {
      chain.register(computational("s1", [{ kind: "post_turn" }], "pass"));
      expect.fail("expected duplicate registration to throw");
    } catch (e) {
      expect(e).toBeInstanceOf(SensorError);
      expect((e as sensor.SensorError).kind).toBe("already_registered");
    }
  });
});

describe("SensorChain — fire", () => {
  it("runs all matching sensors with no short circuit", async () => {
    const chain = new StandardSensorChain();
    chain.register(computational("pass", [{ kind: "post_turn" }], "pass"));
    chain.register(computational("warn", [{ kind: "post_turn" }], "warn"));
    chain.register(computational("halt", [{ kind: "post_turn" }], "halt"));
    const results = await chain.fire({ kind: "post_turn" }, input("s1"));
    expect(results).toHaveLength(3);
    const outcomes = new Set(results.map((r) => r.outcome));
    expect(outcomes.has("pass")).toBe(true);
    expect(outcomes.has("warn")).toBe(true);
    expect(outcomes.has("halt")).toBe(true);
  });

  it("filters by trigger", async () => {
    const chain = new StandardSensorChain();
    chain.register(computational("post-turn", [{ kind: "post_turn" }], "pass"));
    chain.register(computational("post-session", [{ kind: "post_session" }], "pass"));
    const results = await chain.fire({ kind: "post_turn" }, input("s1"));
    expect(results).toHaveLength(1);
    expect(results[0]!.sensor_id.value).toBe("post-turn");
  });

  it("post_tool wildcard and named matching", async () => {
    const chain = new StandardSensorChain();
    chain.register(computational("any", [{ kind: "post_tool", tool_name: "" }], "pass"));
    chain.register(computational("bash-only", [{ kind: "post_tool", tool_name: "bash" }], "pass"));
    const r1 = await chain.fire({ kind: "post_tool", tool_name: "bash" }, input("s1"));
    expect(r1).toHaveLength(2);
    const r2 = await chain.fire({ kind: "post_tool", tool_name: "edit" }, input("s2"));
    expect(r2).toHaveLength(1);
    expect(r2[0]!.sensor_id.value).toBe("any");
  });

  it("computational sensors ignore turn gating", async () => {
    const chain = new StandardSensorChain();
    // Even with `run_every_n_turns` set, computational fires every time.
    chain.register(computational("c", [{ kind: "post_turn" }], "pass", { run_every_n_turns: 99 }));
    const i = input("s1");
    i.turn_number = 1;
    const r = await chain.fire({ kind: "post_turn" }, i);
    expect(r).toHaveLength(1);
  });

  it("inferential gated by run_every_n_turns", async () => {
    const chain = new StandardSensorChain();
    chain.register(inferential("judge", [{ kind: "post_turn" }], "warn", 3, null));
    const i = input("s1");
    i.turn_number = 1;
    expect(await chain.fire({ kind: "post_turn" }, i)).toHaveLength(0);
    i.turn_number = 2;
    expect(await chain.fire({ kind: "post_turn" }, i)).toHaveLength(0);
    i.turn_number = 3;
    expect(await chain.fire({ kind: "post_turn" }, i)).toHaveLength(1);
    i.turn_number = 6;
    expect(await chain.fire({ kind: "post_turn" }, i)).toHaveLength(1);
  });

  it("inferential gated by run_on_phases", async () => {
    const chain = new StandardSensorChain();
    chain.register(inferential("judge", [{ kind: "post_turn" }], "pass", null, ["execution"]));
    const i = input("s1");
    i.phase = "planning";
    expect(await chain.fire({ kind: "post_turn" }, i)).toHaveLength(0);
    i.phase = "execution";
    expect(await chain.fire({ kind: "post_turn" }, i)).toHaveLength(1);
  });
});

describe("SensorChain — stats", () => {
  it("aggregates outcomes and fire_rate", async () => {
    const chain = new StandardSensorChain();
    chain.register(computational("warner", [{ kind: "post_turn" }], "warn"));
    for (let i = 0; i < 4; i++) {
      await chain.fire({ kind: "post_turn" }, input(`s${i}`));
    }
    const stats = await chain.stats();
    expect(stats).toHaveLength(1);
    const s = stats[0]!;
    expect(s.total_fires).toBe(4);
    expect(s.warn_count).toBe(4);
    expect(s.halt_count).toBe(0);
    expect(s.pass_count).toBe(0);
    expect(Math.abs(s.fire_rate - 1.0)).toBeLessThan(1e-6);
  });

  it("history is recorded in fire order", async () => {
    const chain = new StandardSensorChain();
    chain.register(computational("s1", [{ kind: "post_turn" }], "pass"));
    await chain.fire({ kind: "post_turn" }, input("a"));
    await chain.fire({ kind: "post_turn" }, input("b"));
    const stats = await chain.stats();
    expect(stats[0]!.total_fires).toBe(2);
  });
});

describe("SensorChain — signalQualityReport", () => {
  it("flags AlwaysFiring", async () => {
    const chain = new StandardSensorChain();
    chain.register(
      computational("noisy", [{ kind: "post_turn" }], "warn", {
        low_signal_threshold: { never_fired_after_n_sessions: 100, always_fired_rate: 0.5 },
      }),
    );
    for (let i = 0; i < 5; i++) {
      await chain.fire({ kind: "post_turn" }, input(`s${i}`));
    }
    const flags = await chain.signalQualityReport(5);
    expect(flags.some((f) => f.kind === "always_firing" && f.sensor_id.value === "noisy")).toBe(
      true,
    );
  });

  it("flags NeverFired", async () => {
    const chain = new StandardSensorChain();
    chain.register(
      computational("quiet", [{ kind: "post_session" }], "pass", {
        low_signal_threshold: { never_fired_after_n_sessions: 3, always_fired_rate: 0.9 },
      }),
    );
    // Fire PostTurn many times so chain observes the sessions; quiet never fires.
    for (let i = 0; i < 5; i++) {
      await chain.fire({ kind: "post_turn" }, input(`s${i}`));
    }
    const flags = await chain.signalQualityReport(3);
    expect(
      flags.some(
        (f) =>
          f.kind === "never_fired" && f.sensor_id.value === "quiet" && f.sessions_observed >= 3,
      ),
    ).toBe(true);
  });

  it("returns empty when below min_sessions", async () => {
    const chain = new StandardSensorChain();
    chain.register(computational("quiet", [{ kind: "post_session" }], "pass"));
    await chain.fire({ kind: "post_turn" }, input("s1"));
    const flags = await chain.signalQualityReport(10);
    expect(flags).toHaveLength(0);
  });
});

File: go/spore-core/sensor/sensor_test.go
package sensor

import (
	"context"
	"encoding/json"
	"testing"

	sporecore "github.com/squirrelsoft-dev/spore-core/go/spore-core"
)

// stubSensor is a programmable Sensor used across the test suite.
type stubSensor struct {
	cfg     SensorConfig
	outcome SensorOutcome
}

func (s *stubSensor) Evaluate(_ context.Context, _ *SensorInput) SensorResult {
	var obs *string
	if s.outcome == OutcomeWarn {
		v := "warn-obs"
		obs = &v
	}
	return SensorResult{
		SensorID:    s.cfg.ID,
		Outcome:     s.outcome,
		Observation: obs,
		Detail:      string(s.outcome),
		FiredAt:     "2026-05-16T00:00:00Z",
	}
}

func (s *stubSensor) Config() SensorConfig { return s.cfg }

func computational(id string, triggers []SensorTrigger, outcome SensorOutcome) *stubSensor {
	return &stubSensor{
		cfg: SensorConfig{
			ID:                 SensorID(id),
			Name:               id,
			Kind:               SensorKindComputational,
			Triggers:           triggers,
			LowSignalThreshold: DefaultSensorSignalThresholds(),
		},
		outcome: outcome,
	}
}

func inferential(
	id string,
	triggers []SensorTrigger,
	outcome SensorOutcome,
	everyN *uint32,
	phases []sporecore.TaskPhase,
) *stubSensor {
	return &stubSensor{
		cfg: SensorConfig{
			ID:                 SensorID(id),
			Name:               id,
			Kind:               SensorKindInferential,
			Triggers:           triggers,
			RunEveryNTurns:     everyN,
			RunOnPhases:        phases,
			LowSignalThreshold: DefaultSensorSignalThresholds(),
		},
		outcome: outcome,
	}
}

func input(sid string) *SensorInput {
	in := NewSensorInput(sporecore.SessionID(sid), sporecore.SessionState{})
	return &in
}

func u32(v uint32) *uint32 { return &v }

// ── Rule: register validates triggers ───────────────────────────────────────

func TestRegisterRejectsEmptyTriggers(t *testing.T) {
	chain := NewStandardSensorChain()
	err := chain.Register(context.Background(), computational("s1", nil, OutcomePass))
	if err == nil {
		t.Fatal("expected error")
	}
	se, ok := err.(*SensorError)
	if !ok || se.Kind != ErrKindValidationFailed {
		t.Fatalf("expected ValidationFailed, got %v", err)
	}
}

func TestRegisterRejectsDuplicateIDs(t *testing.T) {
	chain := NewStandardSensorChain()
	if err := chain.Register(context.Background(), computational("s1",
		[]SensorTrigger{NewTriggerPostTurn()}, OutcomePass)); err != nil {
		t.Fatal(err)
	}
	err := chain.Register(context.Background(), computational("s1",
		[]SensorTrigger{NewTriggerPostTurn()}, OutcomePass))
	se, ok := err.(*SensorError)
	if !ok || se.Kind != ErrKindAlreadyRegistered {
		t.Fatalf("expected AlreadyRegistered, got %v", err)
	}
}

// ── Rule: fire runs every matching sensor, returns all results ──────────────

func TestFireRunsAllMatchingSensorsNoShortCircuit(t *testing.T) {
	chain := NewStandardSensorChain()
	for _, o := range []SensorOutcome{OutcomePass, OutcomeWarn, OutcomeHalt} {
		if err := chain.Register(context.Background(), computational(string(o),
			[]SensorTrigger{NewTriggerPostTurn()}, o)); err != nil {
			t.Fatal(err)
		}
	}
	results := chain.Fire(context.Background(), NewTriggerPostTurn(), input("s1"))
	if len(results) != 3 {
		t.Fatalf("expected 3 results, got %d", len(results))
	}
	seen := map[SensorOutcome]bool{}
	for _, r := range results {
		seen[r.Outcome] = true
	}
	for _, o := range []SensorOutcome{OutcomePass, OutcomeWarn, OutcomeHalt} {
		if !seen[o] {
			t.Errorf("missing outcome %q", o)
		}
	}
}

// ── Rule: triggers filter ───────────────────────────────────────────────────

func TestFireIgnoresSensorsWithoutMatchingTrigger(t *testing.T) {
	chain := NewStandardSensorChain()
	_ = chain.Register(context.Background(), computational("post-turn",
		[]SensorTrigger{NewTriggerPostTurn()}, OutcomePass))
	_ = chain.Register(context.Background(), computational("post-session",
		[]SensorTrigger{NewTriggerPostSession()}, OutcomePass))
	results := chain.Fire(context.Background(), NewTriggerPostTurn(), input("s1"))
	if len(results) != 1 || results[0].SensorID != "post-turn" {
		t.Fatalf("expected 1 result for post-turn, got %+v", results)
	}
}

// ── Rule: PostTool wildcard / named matching ────────────────────────────────

func TestPostToolWildcardAndNamedMatching(t *testing.T) {
	chain := NewStandardSensorChain()
	_ = chain.Register(context.Background(), computational("any",
		[]SensorTrigger{NewTriggerPostTool("")}, OutcomePass))
	_ = chain.Register(context.Background(), computational("bash-only",
		[]SensorTrigger{NewTriggerPostTool("bash")}, OutcomePass))

	r1 := chain.Fire(context.Background(), NewTriggerPostTool("bash"), input("s1"))
	if len(r1) != 2 {
		t.Fatalf("expected 2 results for bash, got %d", len(r1))
	}
	r2 := chain.Fire(context.Background(), NewTriggerPostTool("edit"), input("s2"))
	if len(r2) != 1 || r2[0].SensorID != "any" {
		t.Fatalf("expected only 'any' for edit, got %+v", r2)
	}
}

// ── Rule: computational ignores RunEveryNTurns ──────────────────────────────

func TestComputationalIgnoresTurnGating(t *testing.T) {
	chain := NewStandardSensorChain()
	s := computational("c", []SensorTrigger{NewTriggerPostTurn()}, OutcomePass)
	s.cfg.RunEveryNTurns = u32(99)
	if err := chain.Register(context.Background(), s); err != nil {
		t.Fatal(err)
	}
	in := input("s1")
	tn := uint32(1)
	in.TurnNumber = &tn
	if r := chain.Fire(context.Background(), NewTriggerPostTurn(), in); len(r) != 1 {
		t.Fatalf("expected computational to fire, got %d", len(r))
	}
}

// ── Rule: inferential gated by RunEveryNTurns ───────────────────────────────

func TestInferentialRunEveryNTurnsGating(t *testing.T) {
	chain := NewStandardSensorChain()
	_ = chain.Register(context.Background(), inferential(
		"judge",
		[]SensorTrigger{NewTriggerPostTurn()},
		OutcomeWarn,
		u32(3),
		nil,
	))
	in := input("s1")
	for _, tc := range []struct {
		turn   uint32
		expect int
	}{
		{1, 0},
		{2, 0},
		{3, 1},
		{6, 1},
	} {
		tn := tc.turn
		in.TurnNumber = &tn
		r := chain.Fire(context.Background(), NewTriggerPostTurn(), in)
		if len(r) != tc.expect {
			t.Errorf("turn %d: expected %d, got %d", tc.turn, tc.expect, len(r))
		}
	}
}

// ── Rule: inferential gated by RunOnPhases ──────────────────────────────────

func TestInferentialRunOnPhasesGating(t *testing.T) {
	chain := NewStandardSensorChain()
	_ = chain.Register(context.Background(), inferential(
		"judge",
		[]SensorTrigger{NewTriggerPostTurn()},
		OutcomePass,
		nil,
		[]sporecore.TaskPhase{sporecore.PhaseExecution},
	))
	in := input("s1")
	planning := sporecore.PhasePlanning
	in.Phase = &planning
	if r := chain.Fire(context.Background(), NewTriggerPostTurn(), in); len(r) != 0 {
		t.Fatalf("expected 0 when phase=Planning, got %d", len(r))
	}
	execution := sporecore.PhaseExecution
	in.Phase = &execution
	if r := chain.Fire(context.Background(), NewTriggerPostTurn(), in); len(r) != 1 {
		t.Fatalf("expected 1 when phase=Execution, got %d", len(r))
	}
}

// ── Rule: stats aggregates outcomes and fire_rate ───────────────────────────

func TestStatsAggregatesOutcomesAndFireRate(t *testing.T) {
	chain := NewStandardSensorChain()
	_ = chain.Register(context.Background(), computational("warner",
		[]SensorTrigger{NewTriggerPostTurn()}, OutcomeWarn))
	for i := 0; i < 4; i++ {
		chain.Fire(context.Background(), NewTriggerPostTurn(), input(string(rune('a'+i))))
	}
	stats := chain.Stats(context.Background(), nil)
	if len(stats) != 1 {
		t.Fatalf("expected 1 stat, got %d", len(stats))
	}
	s := stats[0]
	if s.TotalFires != 4 || s.WarnCount != 4 || s.HaltCount != 0 || s.PassCount != 0 {
		t.Fatalf("aggregation off: %+v", s)
	}
	if d := s.FireRate - 1.0; d < -1e-6 || d > 1e-6 {
		t.Fatalf("fire_rate expected ~1.0, got %f", s.FireRate)
	}
}

// ── Rule: SignalQualityReport flags AlwaysFiring ────────────────────────────

func TestSignalQualityFlagsAlwaysFiring(t *testing.T) {
	chain := NewStandardSensorChain()
	s := computational("noisy", []SensorTrigger{NewTriggerPostTurn()}, OutcomeWarn)
	s.cfg.LowSignalThreshold = SensorSignalThresholds{
		NeverFiredAfterNSessions: 100,
		AlwaysFiredRate:          0.5,
	}
	_ = chain.Register(context.Background(), s)
	for i := 0; i < 5; i++ {
		chain.Fire(context.Background(), NewTriggerPostTurn(), input(string(rune('a'+i))))
	}
	flags := chain.SignalQualityReport(context.Background(), 5)
	found := false
	for _, f := range flags {
		if f.Kind == FlagKindAlwaysFiring && f.SensorID == "noisy" {
			found = true
		}
	}
	if !found {
		t.Fatalf("expected AlwaysFiring flag for 'noisy', got %+v", flags)
	}
}

// ── Rule: SignalQualityReport flags NeverFired ──────────────────────────────

func TestSignalQualityFlagsNeverFired(t *testing.T) {
	chain := NewStandardSensorChain()
	s := computational("quiet", []SensorTrigger{NewTriggerPostSession()}, OutcomePass)
	s.cfg.LowSignalThreshold = SensorSignalThresholds{
		NeverFiredAfterNSessions: 3,
		AlwaysFiredRate:          0.9,
	}
	_ = chain.Register(context.Background(), s)
	for i := 0; i < 5; i++ {
		chain.Fire(context.Background(), NewTriggerPostTurn(), input(string(rune('a'+i))))
	}
	flags := chain.SignalQualityReport(context.Background(), 3)
	found := false
	for _, f := range flags {
		if f.Kind == FlagKindNeverFired && f.SensorID == "quiet" && f.SessionsObserved >= 3 {
			found = true
		}
	}
	if !found {
		t.Fatalf("expected NeverFired flag for 'quiet', got %+v", flags)
	}
}

// ── Rule: SignalQualityReport respects min_sessions ─────────────────────────

func TestSignalQualityRespectsMinSessions(t *testing.T) {
	chain := NewStandardSensorChain()
	_ = chain.Register(context.Background(), computational("quiet",
		[]SensorTrigger{NewTriggerPostSession()}, OutcomePass))
	chain.Fire(context.Background(), NewTriggerPostTurn(), input("s1"))
	flags := chain.SignalQualityReport(context.Background(), 10)
	if len(flags) != 0 {
		t.Fatalf("expected empty flags below min_sessions, got %+v", flags)
	}
}

// ── Edge: history records every fire ────────────────────────────────────────

func TestHistoryIsRecordedInFireOrder(t *testing.T) {
	chain := NewStandardSensorChain()
	_ = chain.Register(context.Background(), computational("s1",
		[]SensorTrigger{NewTriggerPostTurn()}, OutcomePass))
	chain.Fire(context.Background(), NewTriggerPostTurn(), input("a"))
	chain.Fire(context.Background(), NewTriggerPostTurn(), input("b"))
	stats := chain.Stats(context.Background(), nil)
	if stats[0].TotalFires != 2 {
		t.Fatalf("expected 2 fires, got %d", stats[0].TotalFires)
	}
}

// ── Trigger JSON round-trip ─────────────────────────────────────────────────

func TestSensorTriggerJSONRoundTrip(t *testing.T) {
	cases := []SensorTrigger{
		NewTriggerPostTool("bash"),
		NewTriggerPostTool(""),
		NewTriggerPostTurn(),
		NewTriggerPostSession(),
		NewTriggerContinuous(),
		NewTriggerOnToolError(),
		NewTriggerOnCompaction(),
	}
	for _, c := range cases {
		b, err := json.Marshal(c)
		if err != nil {
			t.Fatal(err)
		}
		var got SensorTrigger
		if err := json.Unmarshal(b, &got); err != nil {
			t.Fatal(err)
		}
		if got != c {
			t.Errorf("round-trip mismatch: got %+v want %+v (raw=%s)", got, c, b)
		}
	}
}
